### DISCUSSION TITLE
Max batteries per box question

### DISCUSSION THREAD

**Discussion by gurji999** [dvdoug/BoxPacker#598] [General]
Hey, I am trying to implement this. I have everything working the way I want, however, I can't quite figure out how to implement this: https://boxpacker.io/en/stable/custom-constraints.html

I run the packer with all the items in the shopping cart, and it returns the boxes with each of the items placed in their boxes.  However, if I have batteries, I don't want more than 2 per box.  I need the system to spread them across the boxes, and even add a box if needed.

But I am not sure exactly how to implement this.  Keep in mind, I am relatively new with PHP, I have a good knowledge of it, but I haven't quite grasped how classes work yet.  So please dumb it down for me a little.

This is a sample of what I am using right now.  It is based on your docs.
`<?php

header("Content-Type:application/json; charset=utf-8");
require_once('vendor/autoload.php');

use DVDoug\BoxPacker\Rotation;
use DVDoug\BoxPacker\Packer;
use DVDoug\BoxPacker\Test\TestBox;  // use your own `Box` implementation
use DVDoug\BoxPacker\Test\TestItem; // use your own `Item` implementation

##################################################################

	$packer = new Packer();
	$useWeightDistribution = 1;
	if ($useWeightDistribution !== 1)		{
		$packer->setMaxBoxesToBalanceWeight(0);
	}

$packer->addBox(
	new TestBox(
		reference: 'Le petite box',
		outerWidth: 300,
		outerLength: 300,
		outerDepth: 10,
		emptyWeight: 10,
		innerWidth: 296,
		innerLength: 296,
		innerDepth: 8,
		maxWeight: 50
	)
);
$packer->addBox(
	new TestBox(
		reference: 'Le grande box',
		outerWidth: 3000,
		outerLength: 3000,
		outerDepth: 100,
		emptyWeight: 10,
		innerWidth: 2960,
		innerLength: 2960,
		innerDepth: 80,
		maxWeight: 50
	)
);

$packer->addItem(
	item: new TestItem(
		description: 'Item 1',
		width: 0,
		length: 0,
		depth: 0,
		weight: 27,
		allowedRotation: Rotation::BestFit
	),
	qty: 2
);
$packer->addItem(
	item: new TestItem(
		description: 'Item 2',
		width: 0,
		length: 0,
		depth: 0,
		weight: 28,
		allowedRotation: Rotation::BestFit
	),
	qty: 2
);
$packer->addItem(
	item: new TestItem(
		description: 'Item 3',
		width: 0,
		length: 0,
		depth: 0,
		weight: 26,
		allowedRotation: Rotation::BestFit
	),
	qty: 1
);

###############################################

try {
	$packedBoxes = $packer->pack();

} catch (NoBoxesAvailableException $e) {
	$affectedItems = $e->getAffectedItems();
}

print_r($affectedItems);

########################HOW IT WAS PACKED########################################

	echo PHP_EOL;echo PHP_EOL;
	foreach ($packedBoxes as $packedBox) {
		$packedItems = $packedBox->items;
		foreach ($packedItems as $packedItem) { // $packedItem->item is your own item object
			echo $packedItem->item->getDescription() .  ' was packed at co-ordinate ' ;
			echo '(' . $packedItem->x . ', ' . $packedItem->y . ', ' . $packedItem->z . ') with ';
			echo 'w' . $packedItem->width . ', l' . $packedItem->length . ', d' . $packedItem->depth;
			echo PHP_EOL;
		}
	}

echo "These items fitted into " . count($packedBoxes) . " box(es)" . PHP_EOL;

print_r($packedBoxes);

`

**Comment by dvdoug** [comment]
Hi @gurji999 

If you're still working out classes, this is probably going to be a little tough for you - the first question I would have is are you *only* packing batteries, or are you packing a mixture of batteries and not-batteries, as that will influence things a bit.

BoxPacker works with `Item` and `Box` interfaces - an interface is like a "template" for a class without being one itself. The examples in the getting started sections of the docs use the `Test` implementations from the test suite so that copy/pasting the examples is actually runnable, but they're not guaranteed to never-change, which is why all the examples say "use your own implementation". But we can leave that for now.

To make the battery example work, it's not enough to use the `Item` interface (template), there's a more specialised one that must be used `ConstrainedPlacementItem` - the examples you've pasted above (btw use 3 backticks to start an end end code block, that works better with multi-lines) are using `TestItem` which doesn't know anything about constraints. There's a `ConstrainedPlacementByCountTestItem` in the test suite that you could use for the purposes of getting "something" running, there's an example of using it at https://github.com/dvdoug/BoxPacker/blob/master/tests/PackerTest.php#L549. That particular implementation allows you \<x\> number of any particular `description`, if you have more than that it would force a new box. If you have different types of battery with different descriptions then that might not the right check, as you might end up with e.g. 4 different *types* of battery in the same box. In the battery example in the documentation the check is slightly different - it looks at the class type rather than the description to figure out it's a battery, or maybe your items have a flag or attribute on them you could look for.

You can write any kind of check you need, this is why BoxPacker works on the interfaces (templates) and not the actual classes, this gives people like you the ability to write the custom code you need to handle your particular scenarios. Only you know whether your particular item is a battery or has any other kind of hazard, and only you know how many can go into a box.

**Reply by gurji999** [reply]
Hi, I appreciate the in depth answer!  Let take a look at what you wrote here and I will follow up.  
Thanks!

**Reply by gurji999** [reply]
Ok, I got it to do what I needed.  Probably not the way you would have done it, but this is how I learn. Trial and error and messing around till I get it.

So it lets me control how many batteries per box while also letting unlimited non-batteries in the same boxes.

So my next question would be, what if I didn't want anything in the box besides 2 batteries per box.  Other items would need a separate box.




I really appreciate your excellent work on this project and support!

**Reply by dvdoug** [reply]
It would be something like this (haven't tested it, might have a bug)

```
       public function canBePacked(
        PackedBox $packedBox,
        int $proposedX,
        int $proposedY,
        int $proposedZ,
        int $width,
        int $length,
        int $depth
    ): bool {
        $batteryCount = 0;
        $notBatteryCount = 0;
        $thisItemIsABattery = ($this->getDescription() === 'Battery'); // or whatever your check is

        foreach ($packedBox->items as $alreadyPackedItem) {
            if ($alreadyPackedItem->item->getDescription() === 'Battery') { // or whatever your check is
                $batteryCount++;
            } else {
                $notBatteryCount++;
            }
        }

        if ($thisItemIsABattery) { // can't share with any non-batteries, also max count to be enforced
            return $notBatteryCount === 0 && ($batteryCount + 1) <= static::$limit;
        } elseif ($batteryCount > 0) {
            return false; // not a battery, but can't share with batteries
        } else {
            return true; // not a battery, also no batteries in box
        }
    }
```

**Reply by gurji999** [reply]
Excellent!  Thank you again for this.

**Comment by gurji999** [comment]
Following up on this.  Is it possible to have multiple constrained items but with different limits?
Example:
max item A: 2.
max item B 4.
